CompilerSource/parsing/macros.cpp

```diff
@@ -0,0 +1,316 @@
+    while (begin != end && begin->PreprocessesAway()) ++begin;
+    if (!dest.empty() && begin != end) {
+      Token &left = dest.back();
+      left = PasteTokens(left, *begin++, errc);
+    }
+  }
+  dest.insert(dest.end(), begin, end);
+}
+
+TokenVector Macro::SubstituteAndUnroll(
+    const vector<TokenVector> &args, const vector<TokenVector> &args_evald,
+    ErrorContext errc) const {
+  TokenVector res;
+  bool paste_next = false;
+  if (args.size() < parameters->size()) {
+    if (!is_variadic || args.size() + 1 < parameters->size()) {
+      errc.Error() << "Too few arguments to macro " << NameAndPrototype()
+                   << ": wanted " << parameters->size()
+                   << ", got " << args.size();
+    }
+  }
+  if (args.size() > parameters->size()) {
+    if (!is_variadic) {
+      errc.Error() << "Too many arguments to macro " << NameAndPrototype()
+                   << ": wanted " << parameters->size()
+                   << ", got " << args.size();
+    } else {
+      errc.Error() << "Internal error: variadic macro passed too many arguments";
+    }
+  }
+  // Errors from here on out will concern tokens.
+  for (const FuncComponent &part : parts) {
+    switch (part.tag) {
+      case FuncComponent::TOKEN_SPAN: {
+        AppendOrPaste(res, value.begin() + part.token_span.begin,
+                           value.begin() + part.token_span.end,
+                      paste_next, errc);
+        paste_next = false;
+        break;
+      }
+      case FuncComponent::RAW_ARGUMENT:
+      case FuncComponent::EXPANDED_ARGUMENT:
+      case FuncComponent::STRINGIFY: {
+        size_t ind = part.raw_expanded_or_stringify_argument.index;
+        if (ind >= args.size()) {
+          if (ind >= parameters->size()) {
+            errc.Error() << "Internal error: "
+                << "Macro function built with bad argument references. Index "
+                << ind << " out of bounds (only " << parameters->size()
+                << " params defined).";
+            paste_next = false;
+            continue;
+          } else {
+            paste_next = false;
+            continue;
+          }
+        }
+        if (part.tag == FuncComponent::STRINGIFY) {
+          string str;
+          for (const Token &tok : args[ind])
+            str += tok.content;
+          str = QuoteString(str);
+          TokenVector vec{Token(TT_STRING, errc.snippet)};
+          AppendOrPaste(res, vec.begin(), vec.end(), paste_next, errc);
+          paste_next = false;
+          break;
+        }
+        const TokenVector &vec =
+            part.tag == FuncComponent::EXPANDED_ARGUMENT
+                ? args_evald[ind] : args[ind];
+        AppendOrPaste(res, vec.begin(), vec.end(), paste_next, errc);
+        paste_next = false;
+        break;
+      }
+      case FuncComponent::PASTE:
+        paste_next = true;
+        break;
+      case FuncComponent::VA_OPT: {
```

Comment:
It might be helpful to have some comments on what each of the switch cases are used for exactly